# Part I: Multiple Choice Questions

#### 1. (5 points) If our regression equation is $y = X\beta + u$, where we have $T$ observations and $K$ regressors, what will be the dimension of $\beta$ using the standard matrix notation

> (a) $T \times K$
>
> (b) $T \times 1$
>
> (c) $K \times 1$
>
> (d) $K \times K.$


**Your answer:**

C



#### 2. (5 points) Suppose that the value of $R^2$ for an estimated regression model is exactly one. Which of the following are true?

>> i. All of the data points must lie exactly on the line
>>
>> ii. All of the residuals must be zero
>>
>> iii. All of the variability of $y$ about its mean has been explained by the model
>>
>> iv. The fitted line will be horizontal with respect to all of the explanatory variables.
>
> (a) (ii) and (iv) only
>
> (b) (i) and (iii) only
>
> (c) (i), (ii), and (iii) only
>
> (d) (i), (ii), (iii), and (iv).


**Your answer:**

C

#### 3. (5 points) Which of these is a mathematical expression of the residual sum of squares?

>> i. $\hat{u}^\prime \hat{u}$
>>
>> ii. $\left[\hat{u}_1, \hat{u}_2, ..., \hat{u}_T \right]$
>>
>> iii. $\hat{u}_1 + \hat{u}_2 + ... + \hat{u}_T$
>
> (a) (i) only
>
> (b) (i) and (ii) only
>
> (c) (i) and (iii) only
>
> (d) (i), (ii), and (iii).


**Your answer:**

A


#### 4. (5 points) Why is $R^2$ a commonly used and perhaps better measure of how well a regression model fits the data than the residual sum of squares (RSS)?

> (a) The RSS is often too large
>
> (b) The RSS does not depend on the scale of the dependent variable whereas the $R^2$ does
>
> (c) The RSS depends on the scale of the dependent variable whereas the $R^2$  does not
>
> (d) The RSS depends on the scale of the independent variable whereas the $R^2$ does not.


**Your answer:**

C

#### 5. (5 points) In the following regression estimated on 64 observations
        
$$y_t = \beta_1 + \beta_2 X_{2,t} + \beta_3 X_{3,t} + \beta_4 X_{4,t} + u_t,$$

**Which of the following null hypotheses could we test using an $F$-test?**

>> i. $\beta_2 = 0$
>>
>> ii. $\beta_2 = 1$ and $\beta_3 + \beta_4 = 1$
>>
>> iii. $\beta_3 \beta_4 = 1$
>>
>> iv. $\beta_2 - \beta_3 - \beta_4 = 1.$
>
> (a) (i) and (ii) only
>
> (b) (ii) and (iv) only
>
> (c) (i) and (iii) only
>
> (d) (i), (ii), (iii), and (iv)
>
> (e) (i), (ii), (iv) only.
>


**Your answer:**

E

# Part II: The Fama and French Model

## The Framework
We briefly explore the mathematical and explanatory description of key asset pricing models (i.e., CAPM, Fama-French 3 Factor, Fama-French 5 factor), and how to run these models in Python.

### Capital Asset Pricing Model (CAPM)
We have already introduced and estimated the Capital Asset Pricing Model (CAPM). The CAPM model explains the relationship between systematic risk and the expected return for assets (e.g., stocks). The CAPM is used for pricing of risky assets, by generating the expected return of the asset given its riskiness, and calculating the cost of capital. Intuitively, the CAPM model tells us that the return of a risky asset is explained by the market factor (i.e., $r_m$)

$$\bar{r}_a = r_f + \beta_a (\bar{r}_m - r_f)$$


$\bar{r}_a$: The return of the asset. This is can be the return of the any stock (i.e., Apple, Google, Tesla) or investment portfolio (i.e., any mutual/hedge fund portfolio).

$r_f$: The return of the risk-free asset. The risk-free asset is usually given by the US 3-month Treasury bill. It is assumed that the US government will not default on a short-term government security, thus the US 3-month Treasury bill is widely assumed in finance to be risk-free. $\bar{r}_m$

$\bar{r}_m$​​ : The return of the market. This is usually given by the S&P500 return as it is the largest market index in the world.

\\
The CAPM can also be expressed mathematically in the following notation:
$$r-r_f = \alpha + \beta_M (MKT-r_f)$$


We will use the above notation going forward.

Note: If you are running an asset pricing model on stocks in countries outside the US, it may be more applicable to change $r_m$ ($r_f$) to the market index (short-term government security) of that country.

### Fama-French 3-factor (FF3)
Another very popular asset pricing model in the empirical finance literature is the Fama-French 3-factor (FF3) that was published in 1993. Nobel Laureate Eugene Fama and researcher Kenneth French found that value stocks tend to outperform growth stocks (i.e., value), and that small-cap stocks outperform large-cap stocks (i.e., size). Thus, the FF3 mode adds in size and value as risk factors to the model as shown below

$$r-r_f = \alpha + \beta_M (MKT-r_f) + \beta_S SMB + \beta_v HML$$

### Fama-French 5-factor (FF5)
In 2015, Fama-French added two more risk factors into their popular 3-factor asset pricing model to make a Fama-French 5-factor (FF5) model. This model added two 'quality' factors, namely profitability (stocks with a high operating profitability perform better) and investment (stocks of companies with high total asset growth have below average returns) factors.

$$r-r_f = \alpha + \beta_M (MKT-r_f) + \beta_S SMB + \beta_v HML + \beta_r RMW + \beta_c CMA$$

SMB: The return spread of small minus large stocks (size).

HML: The return of cheap minus expensive stocks (value).

RMW: The return spread of the most profitable firms minus the least profitable (profit).

CMA: The return spread of firms that invest conservatively minus aggressively (investment).



## Solution

### Importing the Python modules
When you use the import *module* as *short_form*, this allows you to apply functions with the short_form instead. For example, when using import numpy as np, numpy.log becomes np.log.

In [ ]:
import pandas as pd
import numpy as np

import statsmodels.formula.api as sm # module for stats models
from statsmodels.iolib.summary2 import summary_col # module for presenting stats models outputs nicely

### Defining the Python regression function
The function extracts the risk factor returns from Ken French's website and runs a CAPM, FF3, and FF5 regression. Input: Pandas series of a stock/portfolio returns.

In [ ]:
def assetPriceReg(df_stk):
  import pandas_datareader.data as web  # module for reading datasets directly from the web

  # Reading in factor data
  df_factors = web.DataReader('F-F_Research_Data_5_Factors_2x3', 'famafrench', start=1900)[0]
  df_factors.rename(columns={'Mkt-RF': 'MKT'}, inplace=True)
  df_factors['MKT'] = df_factors['MKT']/100
  df_factors['SMB'] = df_factors['SMB']/100
  df_factors['HML'] = df_factors['HML']/100
  df_factors['RF'] = df_factors['RF']/100

  df_stk_name = df_stk.name
  df_stock_factor = pd.merge(df_stk,df_factors,left_index=True,right_index=True) # Merging the stock and factor returns dataframes together
  df_stock_factor['XsRet'] = df_stock_factor[df_stk_name] - df_stock_factor['RF'] # Calculating excess returns

  # Running CAPM, FF3, and FF5 models.
  CAPM = sm.ols(formula = 'XsRet ~ MKT', data=df_stock_factor).fit(cov_type='HAC',cov_kwds={'maxlags':1})
  FF3 = sm.ols( formula = 'XsRet ~ MKT + SMB + HML', data=df_stock_factor).fit(cov_type='HAC',cov_kwds={'maxlags':1})

  CAPMtstat = CAPM.tvalues
  FF3tstat = FF3.tvalues

  CAPMcoeff = CAPM.params
  FF3coeff = FF3.params

  # DataFrame with coefficients and t-stats
  results_df = pd.DataFrame({'CAPMcoeff':CAPMcoeff,'CAPMtstat':CAPMtstat,
                              'FF3coeff':FF3coeff, 'FF3tstat':FF3tstat},
  index = ['Intercept', 'MKT', 'SMB', 'HML'])


  dfoutput = summary_col([CAPM,FF3],stars=True,float_format='%0.4f',
                model_names=['CAPM','FF3'],
                info_dict={'N':lambda x: "{0:d}".format(int(x.nobs)),
                            'Adjusted R2':lambda x: "{:.4f}".format(x.rsquared_adj)},
                            regressor_order = ['Intercept', 'MKT', 'SMB', 'HML', 'RMW', 'CMA'])

  print(dfoutput)

  return {'results': results_df, 'CAPM': CAPM, 'FF3': FF3}

### Loading portfolios from Ken French's website
First, let's explore available datasets on Ken French's website

In [ ]:
from pandas_datareader.famafrench import get_available_datasets
datasets = get_available_datasets()
print(datasets)

['F-F_Research_Data_Factors', 'F-F_Research_Data_Factors_weekly', 'F-F_Research_Data_Factors_daily', 'F-F_Research_Data_5_Factors_2x3', 'F-F_Research_Data_5_Factors_2x3_daily', 'Portfolios_Formed_on_ME', 'Portfolios_Formed_on_ME_Wout_Div', 'Portfolios_Formed_on_ME_Daily', 'Portfolios_Formed_on_BE-ME', 'Portfolios_Formed_on_BE-ME_Wout_Div', 'Portfolios_Formed_on_BE-ME_Daily', 'Portfolios_Formed_on_OP', 'Portfolios_Formed_on_OP_Wout_Div', 'Portfolios_Formed_on_OP_Daily', 'Portfolios_Formed_on_INV', 'Portfolios_Formed_on_INV_Wout_Div', 'Portfolios_Formed_on_INV_Daily', '6_Portfolios_2x3', '6_Portfolios_2x3_Wout_Div', '6_Portfolios_2x3_weekly', '6_Portfolios_2x3_daily', '25_Portfolios_5x5', '25_Portfolios_5x5_Wout_Div', '25_Portfolios_5x5_Daily', '100_Portfolios_10x10', '100_Portfolios_10x10_Wout_Div', '100_Portfolios_10x10_Daily', '6_Portfolios_ME_OP_2x3', '6_Portfolios_ME_OP_2x3_Wout_Div', '6_Portfolios_ME_OP_2x3_daily', '25_Portfolios_ME_OP_5x5', '25_Portfolios_ME_OP_5x5_Wout_Div', '25_

Now, let's import the dataset containing 5x5 portfolios sorted on market equity (ME) and book-to-market ratio (B/M). It's a large file, so be patient.

The file is split into multiple segments; the first one contains daily returns on 25 portfolios.

In [ ]:
import pandas_datareader.data as web  # module for reading datasets directly from the web
r = web.DataReader('25_Portfolios_5x5', 'famafrench', start=1900)[0]/100

In [ ]:
r.head()

SMALL LoBM   ME1 BM2   ME1 BM3   ME1 BM4  SMALL HiBM   ME2 BM1  \
Date                                                                      
1926-07    0.058248 -0.017006  0.004875 -0.014580    0.020534  0.012077   
1926-08   -0.020206 -0.080282  0.013796  0.014606    0.083968  0.023618   
1926-09   -0.048291 -0.026154 -0.043417 -0.032729    0.008649 -0.026540   
1926-10   -0.093729 -0.035519 -0.034948  0.034413   -0.025476 -0.028069   
1926-11    0.055888  0.041877  0.024623 -0.044494    0.005362  0.031033   

          ME2 BM2   ME2 BM3   ME2 BM4   ME2 BM5  ...   ME4 BM1   ME4 BM2  \
Date                                             ...                       
1926-07  0.024192  0.004926 -0.026049 -0.003344  ...  0.015893  0.015278   
1926-08 -0.011849  0.040084  0.005038  0.061675  ...  0.013336  0.038730   
1926-09 -0.012618  0.010829 -0.035480 -0.009401  ...  0.010923 -0.005250   
1926-10 -0.032663 -0.050745 -0.080191 -0.013213  ... -0.033361 -0.026559   
1926-11 -0.023690  0.030078  0.051546  0.027292  ...  0.034448  0.023887   

          ME4 BM3   ME4 BM4   ME4 BM5  BIG LoBM   ME5 BM2   ME5 BM3   ME5 BM4  \
Date                                                                            
1926-07  0.012978  0.002727  0.024678  0.034539  0.060902  0.020266  0.031111   
1926-08  0.020021  0.021706  0.053422  0.010124  0.041903  0.020131  0.054849   
1926-09 -0.017636  0.014646  0.008730 -0.012906  0.036538  0.000950 -0.007487   
1926-10 -0.021070 -0.031051 -0.053525 -0.027413 -0.030071 -0.022437 -0.046719   
1926-11  0.037335  0.049320  0.018213  0.042946  0.025326  0.015204  0.036619   

         BIG HiBM  
Date               
1926-07  0.005623  
1926-08  0.077576  
1926-09 -0.024284  
1926-10 -0.058129  
1926-11  0.025636  

[5 rows x 25 columns]

### Run a regression for each stock
Try to collect all alphas into a list. We did this in the *CAPM* notebook. Do the same for betas on SMB and HML. This would make it easier to see any patterns. If you unsure how to do this -- it's okay -- you can simply look the regression output one-by-one and try to find a pattern.

In [ ]:
reg = assetPriceReg(r[r.columns[0]])["results"]


                  CAPM      FF3    
-----------------------------------
Intercept      -0.0049** -0.0047***
               (0.0019)  (0.0009)  
MKT            1.4203*** 1.0867*** 
               (0.0420)  (0.0244)  
SMB                      1.3906*** 
                         (0.0430)  
HML                      -0.4870***
                         (0.0427)  
R-squared      0.6308    0.9098    
R-squared Adj. 0.6303    0.9094    
N              710       710       
Adjusted R2    0.6303    0.9094    
Standard errors in parentheses.
* p<.1, ** p<.05, ***p<.01


Answer and explain the rest of the questions.

In [ ]:
r.columns.values.reshape(5, 5)

array([['SMALL LoBM', 'ME1 BM2', 'ME1 BM3', 'ME1 BM4', 'SMALL HiBM'],
       ['ME2 BM1', 'ME2 BM2', 'ME2 BM3', 'ME2 BM4', 'ME2 BM5'],
       ['ME3 BM1', 'ME3 BM2', 'ME3 BM3', 'ME3 BM4', 'ME3 BM5'],
       ['ME4 BM1', 'ME4 BM2', 'ME4 BM3', 'ME4 BM4', 'ME4 BM5'],
       ['BIG LoBM', 'ME5 BM2', 'ME5 BM3', 'ME5 BM4', 'BIG HiBM']],
      dtype=object)

In [ ]:
np.set_printoptions(precision=3, suppress=True)
r.mean().values.reshape(5, 5)

array([[0.009, 0.01 , 0.013, 0.014, 0.016],
       [0.009, 0.012, 0.012, 0.013, 0.015],
       [0.01 , 0.012, 0.012, 0.013, 0.014],
       [0.01 , 0.011, 0.011, 0.012, 0.013],
       [0.009, 0.009, 0.01 , 0.009, 0.012]])

In [ ]:
#ME1 BM2	0.000324	0.954457	1.309284	-0.191418	0.926001	6.571513e-01
#ME1 BM3
reg = assetPriceReg(r['SMALL HiBM'])["FF3"]
print(reg.bse[0])


                  CAPM      FF3   
----------------------------------
Intercept      0.0053*** 0.0020***
               (0.0016)  (0.0006) 
MKT            1.0703*** 0.9405***
               (0.0450)  (0.0191) 
SMB                      1.0876***
                         (0.0431) 
HML                      0.5281***
                         (0.0308) 
R-squared      0.5799    0.8996   
R-squared Adj. 0.5793    0.8991   
N              710       710      
Adjusted R2    0.5793    0.8991   
Standard errors in parentheses.
* p<.1, ** p<.05, ***p<.01
0.0006234405924756384


In [ ]:
alphas_FF3 = []
betas_FF3_mkt = []
betas_FF3_smb = []
betas_FF3_hml = []
r_squared_FF3 = []
std_error_FF3 = []
tstat_FF3 = []
p_values_FF3 = []


# iterate over all stocks
for c in r.columns:
  reg = assetPriceReg(r[c])["FF3"]
  alphas_FF3.append(reg.params[0])
  betas_FF3_mkt.append(reg.params[1])
  betas_FF3_smb.append(reg.params[2])
  betas_FF3_hml.append(reg.params[3])
  r_squared_FF3.append(reg.rsquared)
  std_error_FF3.append(reg.bse[0])
  tstat_FF3.append(reg.tvalues[0])
  p_values_FF3.append(reg.pvalues[0])



alphas_FF3 = np.array(alphas_FF3)
alphas_FF3 = pd.DataFrame(alphas_FF3, index = r.columns)
alphas_FF3 = alphas_FF3.rename(columns={0: 'alphas_FF3'})
alphas_FF3

betas_FF3_mkt = np.array(betas_FF3_mkt)
betas_FF3_mkt = pd.DataFrame(betas_FF3_mkt, index = r.columns)
betas_FF3_mkt = betas_FF3_mkt.rename(columns={0: 'betas_FF3_mkt'})


betas_FF3_smb = np.array(betas_FF3_smb)
betas_FF3_smb = pd.DataFrame(betas_FF3_smb, index = r.columns)
betas_FF3_smb = betas_FF3_smb.rename(columns={0: 'betas_FF3_smb'})


betas_FF3_hml = np.array(betas_FF3_hml)
betas_FF3_hml = pd.DataFrame(betas_FF3_hml, index = r.columns)
betas_FF3_hml = betas_FF3_hml.rename(columns={0: 'betas_FF3_hml'})


r_squared_FF3 = np.array(r_squared_FF3)
r_squared_FF3 = pd.DataFrame(r_squared_FF3, index = r.columns)
r_squared_FF3 = r_squared_FF3.rename(columns={0: 'r_squared_FF3'})

std_error_FF3 = np.array(std_error_FF3)
std_error_FF3 = pd.DataFrame(std_error_FF3, index = r.columns)
std_error_FF3 = std_error_FF3.rename(columns={0: 'std_error_FF3'})

tstat_FF3 = np.array(tstat_FF3)
tstat_FF3 = pd.DataFrame(tstat_FF3, index = r.columns)
tstat_FF3 = tstat_FF3.rename(columns={0: 'tstat_FF3'})

p_values_FF3 = np.array(p_values_FF3)
p_values_FF3 = pd.DataFrame(p_values_FF3, index = r.columns)
p_values_FF3 = p_values_FF3.rename(columns={0: 'p_values_FF3'})



                  CAPM      FF3    
-----------------------------------
Intercept      -0.0049** -0.0047***
               (0.0019)  (0.0009)  
MKT            1.4203*** 1.0867*** 
               (0.0420)  (0.0244)  
SMB                      1.3906*** 
                         (0.0430)  
HML                      -0.4870***
                         (0.0427)  
R-squared      0.6308    0.9098    
R-squared Adj. 0.6303    0.9094    
N              710       710       
Adjusted R2    0.6303    0.9094    
Standard errors in parentheses.
* p<.1, ** p<.05, ***p<.01

                  CAPM      FF3    
-----------------------------------
Intercept      0.0012    0.0003    
               (0.0016)  (0.0007)  
MKT            1.2300*** 0.9545*** 
               (0.0404)  (0.0214)  
SMB                      1.3093*** 
                         (0.0578)  
HML                      -0.1914***
                         (0.0453)  
R-squared      0.6250    0.9260    
R-squared Adj. 0.6245    0.9257    
N  

In [ ]:
df_FF3 = pd.concat([alphas_FF3, betas_FF3_mkt, betas_FF3_smb, betas_FF3_hml, r_squared_FF3, std_error_FF3, tstat_FF3, p_values_FF3, ], axis = 1)
df_FF3

alphas_FF3  betas_FF3_mkt  betas_FF3_smb  betas_FF3_hml  \
SMALL LoBM   -0.004704       1.086700       1.390633      -0.487034   
ME1 BM2       0.000324       0.954457       1.309284      -0.191418   
ME1 BM3      -0.000170       0.919855       1.090560       0.124242   
ME1 BM4       0.001609       0.877676       1.074757       0.310933   
SMALL HiBM    0.001995       0.940474       1.087624       0.528105   
ME2 BM1      -0.001726       1.116447       1.034919      -0.513227   
ME2 BM2       0.000244       1.002637       0.922244      -0.029060   
ME2 BM3       0.000723       0.965165       0.766746       0.259917   
ME2 BM4       0.000606       0.945648       0.724750       0.453861   
ME2 BM5       0.000270       1.075760       0.879592       0.655156   
ME3 BM1      -0.000933       1.100459       0.748348      -0.503019   
ME3 BM2       0.000809       1.014816       0.606416       0.049068   
ME3 BM3       0.000038       0.972046       0.449752       0.340659   
ME3 BM4       0.000617       0.989258       0.438831       0.539069   
ME3 BM5       0.000426       1.079582       0.574162       0.742143   
ME4 BM1       0.000993       1.065995       0.395063      -0.453161   
ME4 BM2      -0.000406       1.058373       0.239405       0.150888   
ME4 BM3      -0.000110       1.028001       0.184322       0.395040   
ME4 BM4       0.000439       1.023324       0.236446       0.532249   
ME4 BM5      -0.000501       1.138788       0.310417       0.783694   
BIG LoBM      0.001653       0.985033      -0.236220      -0.313488   
ME5 BM2       0.000062       0.974653      -0.171602       0.098307   
ME5 BM3       0.000036       0.943057      -0.202226       0.323079   
ME5 BM4      -0.002164       1.021065      -0.184676       0.692846   
BIG HiBM     -0.001998       1.126506      -0.116034       0.874245   

            r_squared_FF3  std_error_FF3  tstat_FF3  p_values_FF3  
SMALL LoBM       0.909792       0.000903  -5.209919  1.889232e-07  
ME1 BM2          0.926001       0.000729   0.443850  6.571513e-01  
ME1 BM3          0.954379       0.000489  -0.347692  7.280716e-01  
ME1 BM4          0.950279       0.000516   3.116592  1.829543e-03  
SMALL HiBM       0.899551       0.000623   3.200044  1.374067e-03  
ME2 BM1          0.950539       0.000630  -2.737954  6.182275e-03  
ME2 BM2          0.955136       0.000482   0.506436  6.125507e-01  
ME2 BM3          0.937241       0.000533   1.357207  1.747153e-01  
ME2 BM4          0.946903       0.000466   1.301431  1.931111e-01  
ME2 BM5          0.951475       0.000510   0.528445  5.971902e-01  
ME3 BM1          0.947627       0.000600  -1.556241  1.196507e-01  
ME3 BM2          0.928021       0.000554   1.458669  1.446562e-01  
ME3 BM3          0.909281       0.000576   0.066297  9.471410e-01  
ME3 BM4          0.918441       0.000557   1.107741  2.679735e-01  
ME3 BM5          0.902818       0.000712   0.598837  5.492817e-01  
ME4 BM1          0.932136       0.000595   1.666856  9.554295e-02  
ME4 BM2          0.900957       0.000642  -0.633446  5.264426e-01  
ME4 BM3          0.885653       0.000661  -0.165968  8.681818e-01  
ME4 BM4          0.886585       0.000688   0.637518  5.237874e-01  
ME4 BM5          0.880136       0.000781  -0.641197  5.213949e-01  
BIG LoBM         0.945400       0.000425   3.894440  9.842586e-05  
ME5 BM2          0.897069       0.000550   0.113401  9.097125e-01  
ME5 BM3          0.849984       0.000650   0.055580  9.556761e-01  
ME5 BM4          0.888998       0.000668  -3.237419  1.206160e-03  
BIG HiBM         0.803383       0.000986  -2.025729  4.279254e-02

In [ ]:
def mean_of_betas(df):
  for x in df.columns[1:4]:
    x = df[x].mean()
    print(x)


In [ ]:
mean_of_betas(df_FF3)

1.0162309926811424
0.5421405314214436
0.2145237305349336


In [ ]:
alphas_CAPM = []
betas_CAPM = []
r_squared_CAPM = []

# iterate over all stocks
for c in r.columns:
  reg = assetPriceReg(r[c])["CAPM"]
  alphas_CAPM.append(reg.params[0])
  betas_CAPM.append(reg.params[1])
  r_squared_CAPM.append(reg.rsquared)


alphas_CAPM = np.array(alphas_CAPM)
alphas_CAPM = pd.DataFrame(alphas_CAPM, index = r.columns)
alphas_CAPM = alphas_CAPM.rename(columns={0: 'alphas_CAPM'})

betas_CAPM = np.array(betas_CAPM)
betas_CAPM = pd.DataFrame(betas_CAPM, index = r.columns)
betas_CAPM = betas_CAPM.rename(columns={0: 'betas_CAPM'})


r_squared_CAPM  = np.array(r_squared_CAPM )
r_squared_CAPM = pd.DataFrame(r_squared_CAPM, index = r.columns)
r_squared_CAPM = r_squared_CAPM.rename(columns={0: 'r_squared_CAPM'})




                  CAPM      FF3    
-----------------------------------
Intercept      -0.0049** -0.0047***
               (0.0019)  (0.0009)  
MKT            1.4203*** 1.0867*** 
               (0.0420)  (0.0244)  
SMB                      1.3906*** 
                         (0.0430)  
HML                      -0.4870***
                         (0.0427)  
R-squared      0.6308    0.9098    
R-squared Adj. 0.6303    0.9094    
N              710       710       
Adjusted R2    0.6303    0.9094    
Standard errors in parentheses.
* p<.1, ** p<.05, ***p<.01

                  CAPM      FF3    
-----------------------------------
Intercept      0.0012    0.0003    
               (0.0016)  (0.0007)  
MKT            1.2300*** 0.9545*** 
               (0.0404)  (0.0214)  
SMB                      1.3093*** 
                         (0.0578)  
HML                      -0.1914***
                         (0.0453)  
R-squared      0.6250    0.9260    
R-squared Adj. 0.6245    0.9257    
N  

In [ ]:
df_CAPM = pd.concat([alphas_CAPM, betas_CAPM, r_squared_CAPM], axis = 1)
df_CAPM

alphas_CAPM  betas_CAPM  r_squared_CAPM
SMALL LoBM    -0.004858    1.420329        0.630823
ME1 BM2        0.001190    1.230027        0.624986
ME1 BM3        0.001625    1.108459        0.674180
ME1 BM4        0.004092    1.036350        0.633845
SMALL HiBM     0.005316    1.070251        0.579943
ME2 BM1       -0.002411    1.386489        0.740694
ME2 BM2        0.001255    1.181484        0.760513
ME2 BM3        0.002639    1.072870        0.753101
ME2 BM4        0.003205    1.017417        0.724984
ME2 BM5        0.003818    1.147809        0.681445
ME3 BM1       -0.001928    1.314758        0.795714
ME3 BM2        0.001731    1.122579        0.825477
ME3 BM3        0.001874    1.008068        0.800448
ME3 BM4        0.003191    0.994585        0.757890
ME3 BM5        0.003933    1.081237        0.680625
ME4 BM1       -0.000242    1.206195        0.847063
ME4 BM2        0.000456    1.081941        0.875017
ME4 BM3        0.001609    1.005910        0.819479
ME4 BM4        0.002741    0.991308        0.772101
ME4 BM5        0.002842    1.084501        0.699101
BIG LoBM       0.000180    0.985529        0.884590
ME5 BM2        0.000226    0.927971        0.880787
ME5 BM3        0.001013    0.858145        0.787058
ME5 BM4        0.000234    0.886125        0.700574
BIG HiBM       0.001170    0.978393        0.601097

In [ ]:
df_all = pd.concat([df_CAPM, df_FF3], axis = 1)
df_all

alphas_CAPM  betas_CAPM  r_squared_CAPM  alphas_FF3  \
SMALL LoBM    -0.004858    1.420329        0.630823   -0.004704   
ME1 BM2        0.001190    1.230027        0.624986    0.000324   
ME1 BM3        0.001625    1.108459        0.674180   -0.000170   
ME1 BM4        0.004092    1.036350        0.633845    0.001609   
SMALL HiBM     0.005316    1.070251        0.579943    0.001995   
ME2 BM1       -0.002411    1.386489        0.740694   -0.001726   
ME2 BM2        0.001255    1.181484        0.760513    0.000244   
ME2 BM3        0.002639    1.072870        0.753101    0.000723   
ME2 BM4        0.003205    1.017417        0.724984    0.000606   
ME2 BM5        0.003818    1.147809        0.681445    0.000270   
ME3 BM1       -0.001928    1.314758        0.795714   -0.000933   
ME3 BM2        0.001731    1.122579        0.825477    0.000809   
ME3 BM3        0.001874    1.008068        0.800448    0.000038   
ME3 BM4        0.003191    0.994585        0.757890    0.000617   
ME3 BM5        0.003933    1.081237        0.680625    0.000426   
ME4 BM1       -0.000242    1.206195        0.847063    0.000993   
ME4 BM2        0.000456    1.081941        0.875017   -0.000406   
ME4 BM3        0.001609    1.005910        0.819479   -0.000110   
ME4 BM4        0.002741    0.991308        0.772101    0.000439   
ME4 BM5        0.002842    1.084501        0.699101   -0.000501   
BIG LoBM       0.000180    0.985529        0.884590    0.001653   
ME5 BM2        0.000226    0.927971        0.880787    0.000062   
ME5 BM3        0.001013    0.858145        0.787058    0.000036   
ME5 BM4        0.000234    0.886125        0.700574   -0.002164   
BIG HiBM       0.001170    0.978393        0.601097   -0.001998   

            betas_FF3_mkt  betas_FF3_smb  betas_FF3_hml  r_squared_FF3  \
SMALL LoBM       1.086700       1.390633      -0.487034       0.909792   
ME1 BM2          0.954457       1.309284      -0.191418       0.926001   
ME1 BM3          0.919855       1.090560       0.124242       0.954379   
ME1 BM4          0.877676       1.074757       0.310933       0.950279   
SMALL HiBM       0.940474       1.087624       0.528105       0.899551   
ME2 BM1          1.116447       1.034919      -0.513227       0.950539   
ME2 BM2          1.002637       0.922244      -0.029060       0.955136   
ME2 BM3          0.965165       0.766746       0.259917       0.937241   
ME2 BM4          0.945648       0.724750       0.453861       0.946903   
ME2 BM5          1.075760       0.879592       0.655156       0.951475   
ME3 BM1          1.100459       0.748348      -0.503019       0.947627   
ME3 BM2          1.014816       0.606416       0.049068       0.928021   
ME3 BM3          0.972046       0.449752       0.340659       0.909281   
ME3 BM4          0.989258       0.438831       0.539069       0.918441   
ME3 BM5          1.079582       0.574162       0.742143       0.902818   
ME4 BM1          1.065995       0.395063      -0.453161       0.932136   
ME4 BM2          1.058373       0.239405       0.150888       0.900957   
ME4 BM3          1.028001       0.184322       0.395040       0.885653   
ME4 BM4          1.023324       0.236446       0.532249       0.886585   
ME4 BM5          1.138788       0.310417       0.783694       0.880136   
BIG LoBM         0.985033      -0.236220      -0.313488       0.945400   
ME5 BM2          0.974653      -0.171602       0.098307       0.897069   
ME5 BM3          0.943057      -0.202226       0.323079       0.849984   
ME5 BM4          1.021065      -0.184676       0.692846       0.888998   
BIG HiBM         1.126506      -0.116034       0.874245       0.803383   

            std_error_FF3  tstat_FF3  p_values_FF3  
SMALL LoBM       0.000903  -5.209919  1.889232e-07  
ME1 BM2          0.000729   0.443850  6.571513e-01  
ME1 BM3          0.000489  -0.347692  7.280716e-01  
ME1 BM4          0.000516   3.116592  1.829543e-03  
SMALL HiBM       0.000623   3.200044  1.374067e-03  
ME2 BM1          0.000630 

In [ ]:
df_alpha_annual = df_all[["alphas_CAPM", "alphas_FF3"]]*12
df_alpha_annual

alphas_CAPM  alphas_FF3
SMALL LoBM    -0.058292   -0.056451
ME1 BM2        0.014285    0.003885
ME1 BM3        0.019506   -0.002041
ME1 BM4        0.049105    0.019310
SMALL HiBM     0.063787    0.023940
ME2 BM1       -0.028931   -0.020713
ME2 BM2        0.015058    0.002929
ME2 BM3        0.031665    0.008679
ME2 BM4        0.038460    0.007278
ME2 BM5        0.045820    0.003237
ME3 BM1       -0.023135   -0.011202
ME3 BM2        0.020776    0.009704
ME3 BM3        0.022489    0.000459
ME3 BM4        0.038288    0.007405
ME3 BM5        0.047196    0.005116
ME4 BM1       -0.002909    0.011911
ME4 BM2        0.005466   -0.004878
ME4 BM3        0.019314   -0.001316
ME4 BM4        0.032888    0.005266
ME4 BM5        0.034109   -0.006012
BIG LoBM       0.002156    0.019839
ME5 BM2        0.002711    0.000749
ME5 BM3        0.012158    0.000434
ME5 BM4        0.002811   -0.025964
BIG HiBM       0.014042   -0.023972

In [ ]:
reg1 = assetPriceReg(r['SMALL HiBM'])


                  CAPM      FF3   
----------------------------------
Intercept      0.0053*** 0.0020***
               (0.0016)  (0.0006) 
MKT            1.0703*** 0.9405***
               (0.0450)  (0.0191) 
SMB                      1.0876***
                         (0.0431) 
HML                      0.5281***
                         (0.0308) 
R-squared      0.5799    0.8996   
R-squared Adj. 0.5793    0.8991   
N              710       710      
Adjusted R2    0.5793    0.8991   
Standard errors in parentheses.
* p<.1, ** p<.05, ***p<.01


In [ ]:
import pandas_datareader.data as web  # module for reading datasets directly from the web
df_stk = r['SMALL HiBM']
# Reading in factor data
df_factors = web.DataReader('F-F_Research_Data_5_Factors_2x3', 'famafrench', start=1900)[0]
df_factors.rename(columns={'Mkt-RF': 'MKT'}, inplace=True)
df_factors['MKT'] = df_factors['MKT']/100
df_factors['SMB'] = df_factors['SMB']/100
df_factors['HML'] = df_factors['HML']/100
df_factors['RF'] = df_factors['RF']/100

df_stk_name = df_stk.name
df_stock_factor = pd.merge(df_stk,df_factors,left_index=True,right_index=True) # Merging the stock and factor returns dataframes together
df_stock_factor['XsRet'] = df_stock_factor[df_stk_name] - df_stock_factor['RF'] # Calculating excess returns

# Running CAPM, FF3, and FF5 models.
CAPM = sm.ols(formula = 'XsRet ~ MKT', data=df_stock_factor).fit(cov_type='HAC',cov_kwds={'maxlags':1})
FF3 = sm.ols( formula = 'XsRet ~ MKT + SMB + HML', data=df_stock_factor).fit(cov_type='HAC',cov_kwds={'maxlags':1})


In [ ]:
print(FF3.summary())

                            OLS Regression Results                            
Dep. Variable:                  XsRet   R-squared:                       0.900
Model:                            OLS   Adj. R-squared:                  0.899
Method:                 Least Squares   F-statistic:                     2193.
Date:                Sun, 09 Oct 2022   Prob (F-statistic):               0.00
Time:                        22:48:40   Log-Likelihood:                 1772.7
No. Observations:                 710   AIC:                            -3537.
Df Residuals:                     706   BIC:                            -3519.
Df Model:                           3                                         
Covariance Type:                  HAC                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0020      0.001      3.200      0.0

In [ ]:
print(FF3.f_test('SMB = HML'))

<F test: F=array([[137.792]]), p=3.4263493284710105e-29, df_denom=706, df_num=1>


In [ ]:
import scipy.stats as stats
two_sided_tstat = stats.norm.ppf(.975)
two_sided_tstat

1.959963984540054

In [ ]:
negative_side = -two_sided_tstat
negative_side

-1.959963984540054

In [ ]:
df_all['Hypothesis at 5% FF3 T-test'] = df_all['tstat_FF3'].apply(lambda x: 'Reject Null' if x >two_sided_tstat  or x<negative_side else 'Cannot Reject Null')
df_all

alphas_CAPM  betas_CAPM  r_squared_CAPM  alphas_FF3  \
SMALL LoBM    -0.004858    1.420329        0.630823   -0.004704   
ME1 BM2        0.001190    1.230027        0.624986    0.000324   
ME1 BM3        0.001625    1.108459        0.674180   -0.000170   
ME1 BM4        0.004092    1.036350        0.633845    0.001609   
SMALL HiBM     0.005316    1.070251        0.579943    0.001995   
ME2 BM1       -0.002411    1.386489        0.740694   -0.001726   
ME2 BM2        0.001255    1.181484        0.760513    0.000244   
ME2 BM3        0.002639    1.072870        0.753101    0.000723   
ME2 BM4        0.003205    1.017417        0.724984    0.000606   
ME2 BM5        0.003818    1.147809        0.681445    0.000270   
ME3 BM1       -0.001928    1.314758        0.795714   -0.000933   
ME3 BM2        0.001731    1.122579        0.825477    0.000809   
ME3 BM3        0.001874    1.008068        0.800448    0.000038   
ME3 BM4        0.003191    0.994585        0.757890    0.000617   
ME3 BM5        0.003933    1.081237        0.680625    0.000426   
ME4 BM1       -0.000242    1.206195        0.847063    0.000993   
ME4 BM2        0.000456    1.081941        0.875017   -0.000406   
ME4 BM3        0.001609    1.005910        0.819479   -0.000110   
ME4 BM4        0.002741    0.991308        0.772101    0.000439   
ME4 BM5        0.002842    1.084501        0.699101   -0.000501   
BIG LoBM       0.000180    0.985529        0.884590    0.001653   
ME5 BM2        0.000226    0.927971        0.880787    0.000062   
ME5 BM3        0.001013    0.858145        0.787058    0.000036   
ME5 BM4        0.000234    0.886125        0.700574   -0.002164   
BIG HiBM       0.001170    0.978393        0.601097   -0.001998   

            betas_FF3_mkt  betas_FF3_smb  betas_FF3_hml  r_squared_FF3  \
SMALL LoBM       1.086700       1.390633      -0.487034       0.909792   
ME1 BM2          0.954457       1.309284      -0.191418       0.926001   
ME1 BM3          0.919855       1.090560       0.124242       0.954379   
ME1 BM4          0.877676       1.074757       0.310933       0.950279   
SMALL HiBM       0.940474       1.087624       0.528105       0.899551   
ME2 BM1          1.116447       1.034919      -0.513227       0.950539   
ME2 BM2          1.002637       0.922244      -0.029060       0.955136   
ME2 BM3          0.965165       0.766746       0.259917       0.937241   
ME2 BM4          0.945648       0.724750       0.453861       0.946903   
ME2 BM5          1.075760       0.879592       0.655156       0.951475   
ME3 BM1          1.100459       0.748348      -0.503019       0.947627   
ME3 BM2          1.014816       0.606416       0.049068       0.928021   
ME3 BM3          0.972046       0.449752       0.340659       0.909281   
ME3 BM4          0.989258       0.438831       0.539069       0.918441   
ME3 BM5          1.079582       0.574162       0.742143       0.902818   
ME4 BM1          1.065995       0.395063      -0.453161       0.932136   
ME4 BM2          1.058373       0.239405       0.150888       0.900957   
ME4 BM3          1.028001       0.184322       0.395040       0.885653   
ME4 BM4          1.023324       0.236446       0.532249       0.886585   
ME4 BM5          1.138788       0.310417       0.783694       0.880136   
BIG LoBM         0.985033      -0.236220      -0.313488       0.945400   
ME5 BM2          0.974653      -0.171602       0.098307       0.897069   
ME5 BM3          0.943057      -0.202226       0.323079       0.849984   
ME5 BM4          1.021065      -0.184676       0.692846       0.888998   
BIG HiBM         1.126506      -0.116034       0.874245       0.803383   

            std_error_FF3  tstat_FF3  p_values_FF3 Hypothesis at 5% FF3 T-test  
SMALL LoBM       0.000903  -5.209919  1.889232e-07                 Reject Null  
ME1 BM2          0.000729   0.443850  6.571513e-01          Cannot Reject Null  
ME1 BM3          0.000489  -0.347692  7.280716e-01          Cannot Reject Null  
ME1 BM4          0.0

In [ ]:
df_all['Hypothesis at 5% FF3 P-val test'] = df_all['p_values_FF3'].apply(lambda x: 'Reject Null' if x <= 0.05 else 'Cannot Reject Null')
df_all

alphas_CAPM  betas_CAPM  r_squared_CAPM  alphas_FF3  \
SMALL LoBM    -0.004858    1.420329        0.630823   -0.004704   
ME1 BM2        0.001190    1.230027        0.624986    0.000324   
ME1 BM3        0.001625    1.108459        0.674180   -0.000170   
ME1 BM4        0.004092    1.036350        0.633845    0.001609   
SMALL HiBM     0.005316    1.070251        0.579943    0.001995   
ME2 BM1       -0.002411    1.386489        0.740694   -0.001726   
ME2 BM2        0.001255    1.181484        0.760513    0.000244   
ME2 BM3        0.002639    1.072870        0.753101    0.000723   
ME2 BM4        0.003205    1.017417        0.724984    0.000606   
ME2 BM5        0.003818    1.147809        0.681445    0.000270   
ME3 BM1       -0.001928    1.314758        0.795714   -0.000933   
ME3 BM2        0.001731    1.122579        0.825477    0.000809   
ME3 BM3        0.001874    1.008068        0.800448    0.000038   
ME3 BM4        0.003191    0.994585        0.757890    0.000617   
ME3 BM5        0.003933    1.081237        0.680625    0.000426   
ME4 BM1       -0.000242    1.206195        0.847063    0.000993   
ME4 BM2        0.000456    1.081941        0.875017   -0.000406   
ME4 BM3        0.001609    1.005910        0.819479   -0.000110   
ME4 BM4        0.002741    0.991308        0.772101    0.000439   
ME4 BM5        0.002842    1.084501        0.699101   -0.000501   
BIG LoBM       0.000180    0.985529        0.884590    0.001653   
ME5 BM2        0.000226    0.927971        0.880787    0.000062   
ME5 BM3        0.001013    0.858145        0.787058    0.000036   
ME5 BM4        0.000234    0.886125        0.700574   -0.002164   
BIG HiBM       0.001170    0.978393        0.601097   -0.001998   

            betas_FF3_mkt  betas_FF3_smb  betas_FF3_hml  r_squared_FF3  \
SMALL LoBM       1.086700       1.390633      -0.487034       0.909792   
ME1 BM2          0.954457       1.309284      -0.191418       0.926001   
ME1 BM3          0.919855       1.090560       0.124242       0.954379   
ME1 BM4          0.877676       1.074757       0.310933       0.950279   
SMALL HiBM       0.940474       1.087624       0.528105       0.899551   
ME2 BM1          1.116447       1.034919      -0.513227       0.950539   
ME2 BM2          1.002637       0.922244      -0.029060       0.955136   
ME2 BM3          0.965165       0.766746       0.259917       0.937241   
ME2 BM4          0.945648       0.724750       0.453861       0.946903   
ME2 BM5          1.075760       0.879592       0.655156       0.951475   
ME3 BM1          1.100459       0.748348      -0.503019       0.947627   
ME3 BM2          1.014816       0.606416       0.049068       0.928021   
ME3 BM3          0.972046       0.449752       0.340659       0.909281   
ME3 BM4          0.989258       0.438831       0.539069       0.918441   
ME3 BM5          1.079582       0.574162       0.742143       0.902818   
ME4 BM1          1.065995       0.395063      -0.453161       0.932136   
ME4 BM2          1.058373       0.239405       0.150888       0.900957   
ME4 BM3          1.028001       0.184322       0.395040       0.885653   
ME4 BM4          1.023324       0.236446       0.532249       0.886585   
ME4 BM5          1.138788       0.310417       0.783694       0.880136   
BIG LoBM         0.985033      -0.236220      -0.313488       0.945400   
ME5 BM2          0.974653      -0.171602       0.098307       0.897069   
ME5 BM3          0.943057      -0.202226       0.323079       0.849984   
ME5 BM4          1.021065      -0.184676       0.692846       0.888998   
BIG HiBM         1.126506      -0.116034       0.874245       0.803383   

            std_error_FF3  tstat_FF3  p_values_FF3  \
SMALL LoBM       0.000903  -5.209919  1.889232e-07   
ME1 BM2          0.000729   0.443850  6.571513e-01   
ME1 BM3          0.000489  -0.347692  7.280716e-01   
ME1 BM4          0.000516   3.116592  1.829543e-03   
SMALL HiBM       0.000623   3.200044  1.374067e-03   
ME2 BM1          0.0

# **Answer to Questions Part II: The Fama and French Model (40 points):**

### **Question 1)**

Here we can see all the data for the 25 portfolios in a single dataframe for the CAPM and FF3:

In [ ]:
df_all

alphas_CAPM  betas_CAPM  r_squared_CAPM  alphas_FF3  \
SMALL LoBM    -0.004858    1.420329        0.630823   -0.004704   
ME1 BM2        0.001190    1.230027        0.624986    0.000324   
ME1 BM3        0.001625    1.108459        0.674180   -0.000170   
ME1 BM4        0.004092    1.036350        0.633845    0.001609   
SMALL HiBM     0.005316    1.070251        0.579943    0.001995   
ME2 BM1       -0.002411    1.386489        0.740694   -0.001726   
ME2 BM2        0.001255    1.181484        0.760513    0.000244   
ME2 BM3        0.002639    1.072870        0.753101    0.000723   
ME2 BM4        0.003205    1.017417        0.724984    0.000606   
ME2 BM5        0.003818    1.147809        0.681445    0.000270   
ME3 BM1       -0.001928    1.314758        0.795714   -0.000933   
ME3 BM2        0.001731    1.122579        0.825477    0.000809   
ME3 BM3        0.001874    1.008068        0.800448    0.000038   
ME3 BM4        0.003191    0.994585        0.757890    0.000617   
ME3 BM5        0.003933    1.081237        0.680625    0.000426   
ME4 BM1       -0.000242    1.206195        0.847063    0.000993   
ME4 BM2        0.000456    1.081941        0.875017   -0.000406   
ME4 BM3        0.001609    1.005910        0.819479   -0.000110   
ME4 BM4        0.002741    0.991308        0.772101    0.000439   
ME4 BM5        0.002842    1.084501        0.699101   -0.000501   
BIG LoBM       0.000180    0.985529        0.884590    0.001653   
ME5 BM2        0.000226    0.927971        0.880787    0.000062   
ME5 BM3        0.001013    0.858145        0.787058    0.000036   
ME5 BM4        0.000234    0.886125        0.700574   -0.002164   
BIG HiBM       0.001170    0.978393        0.601097   -0.001998   

            betas_FF3_mkt  betas_FF3_smb  betas_FF3_hml  r_squared_FF3  \
SMALL LoBM       1.086700       1.390633      -0.487034       0.909792   
ME1 BM2          0.954457       1.309284      -0.191418       0.926001   
ME1 BM3          0.919855       1.090560       0.124242       0.954379   
ME1 BM4          0.877676       1.074757       0.310933       0.950279   
SMALL HiBM       0.940474       1.087624       0.528105       0.899551   
ME2 BM1          1.116447       1.034919      -0.513227       0.950539   
ME2 BM2          1.002637       0.922244      -0.029060       0.955136   
ME2 BM3          0.965165       0.766746       0.259917       0.937241   
ME2 BM4          0.945648       0.724750       0.453861       0.946903   
ME2 BM5          1.075760       0.879592       0.655156       0.951475   
ME3 BM1          1.100459       0.748348      -0.503019       0.947627   
ME3 BM2          1.014816       0.606416       0.049068       0.928021   
ME3 BM3          0.972046       0.449752       0.340659       0.909281   
ME3 BM4          0.989258       0.438831       0.539069       0.918441   
ME3 BM5          1.079582       0.574162       0.742143       0.902818   
ME4 BM1          1.065995       0.395063      -0.453161       0.932136   
ME4 BM2          1.058373       0.239405       0.150888       0.900957   
ME4 BM3          1.028001       0.184322       0.395040       0.885653   
ME4 BM4          1.023324       0.236446       0.532249       0.886585   
ME4 BM5          1.138788       0.310417       0.783694       0.880136   
BIG LoBM         0.985033      -0.236220      -0.313488       0.945400   
ME5 BM2          0.974653      -0.171602       0.098307       0.897069   
ME5 BM3          0.943057      -0.202226       0.323079       0.849984   
ME5 BM4          1.021065      -0.184676       0.692846       0.888998   
BIG HiBM         1.126506      -0.116034       0.874245       0.803383   

            std_error_FF3  tstat_FF3  p_values_FF3  \
SMALL LoBM       0.000903  -5.209919  1.889232e-07   
ME1 BM2          0.000729   0.443850  6.571513e-01   
ME1 BM3          0.000489  -0.347692  7.280716e-01   
ME1 BM4          0.000516   3.116592  1.829543e-03   
SMALL HiBM       0.000623   3.200044  1.374067e-03   
ME2 BM1          0.0

The folowing dataframe shows annualized alphas for both CAPM and FF3. We can understand that some portfolios have an economically large return and some have a rather low economic return. For example: ME1 BM4 has an annual return of approximately 4.9% which is quite large, and ME4 BM2 has a return of 0.054% which is economically small.
When it comes to the magnitude of the alphas in CAPM and FF3, some portfolios have a close magnitude and others have a high magnitude between the two models.

In [ ]:
df_alpha_annual

alphas_CAPM  alphas_FF3
SMALL LoBM    -0.058292   -0.056451
ME1 BM2        0.014285    0.003885
ME1 BM3        0.019506   -0.002041
ME1 BM4        0.049105    0.019310
SMALL HiBM     0.063787    0.023940
ME2 BM1       -0.028931   -0.020713
ME2 BM2        0.015058    0.002929
ME2 BM3        0.031665    0.008679
ME2 BM4        0.038460    0.007278
ME2 BM5        0.045820    0.003237
ME3 BM1       -0.023135   -0.011202
ME3 BM2        0.020776    0.009704
ME3 BM3        0.022489    0.000459
ME3 BM4        0.038288    0.007405
ME3 BM5        0.047196    0.005116
ME4 BM1       -0.002909    0.011911
ME4 BM2        0.005466   -0.004878
ME4 BM3        0.019314   -0.001316
ME4 BM4        0.032888    0.005266
ME4 BM5        0.034109   -0.006012
BIG LoBM       0.002156    0.019839
ME5 BM2        0.002711    0.000749
ME5 BM3        0.012158    0.000434
ME5 BM4        0.002811   -0.025964
BIG HiBM       0.014042   -0.023972

The R^2 between the CAPM and FF3 differ highly. The R^2 of the FF3 is always higher for each of the 25 portfolios compared to the R^2 of the CAPM. It seems as though the FF3 can better explain the variation of the dependent variables (the 25 portfolios) compared to the CAPM.

In [ ]:
df_all[["r_squared_CAPM", "r_squared_FF3"]]

r_squared_CAPM  r_squared_FF3
SMALL LoBM        0.630823       0.909792
ME1 BM2           0.624986       0.926001
ME1 BM3           0.674180       0.954379
ME1 BM4           0.633845       0.950279
SMALL HiBM        0.579943       0.899551
ME2 BM1           0.740694       0.950539
ME2 BM2           0.760513       0.955136
ME2 BM3           0.753101       0.937241
ME2 BM4           0.724984       0.946903
ME2 BM5           0.681445       0.951475
ME3 BM1           0.795714       0.947627
ME3 BM2           0.825477       0.928021
ME3 BM3           0.800448       0.909281
ME3 BM4           0.757890       0.918441
ME3 BM5           0.680625       0.902818
ME4 BM1           0.847063       0.932136
ME4 BM2           0.875017       0.900957
ME4 BM3           0.819479       0.885653
ME4 BM4           0.772101       0.886585
ME4 BM5           0.699101       0.880136
BIG LoBM          0.884590       0.945400
ME5 BM2           0.880787       0.897069
ME5 BM3           0.787058       0.849984
ME5 BM4           0.700574       0.888998
BIG HiBM          0.601097       0.803383

### Question 2)

Running the F-test for "Small HiBM" where we test the null hypothesis:
βSMB = βHML
From the follwing code we have a very low p_value: 3.43e-29, and so we can conclude that in this case we Reject the Null.

In [ ]:
print(FF3.f_test('SMB = HML'))

<F test: F=array([[137.792]]), p=3.4263493284710105e-29, df_denom=706, df_num=1>


### **Question 3)**

At a 5% level of significance as we can see the following dataframe some portfolios have their null hypothesis rejected and other not rejected by the FF3 model.


In [ ]:
df_all[["std_error_FF3",	"tstat_FF3",	"p_values_FF3",	"Hypothesis at 5% FF3 P-val test",	"Hypothesis at 5% FF3 T-test"]]

std_error_FF3  tstat_FF3  p_values_FF3  \
SMALL LoBM       0.000903  -5.209919  1.889232e-07   
ME1 BM2          0.000729   0.443850  6.571513e-01   
ME1 BM3          0.000489  -0.347692  7.280716e-01   
ME1 BM4          0.000516   3.116592  1.829543e-03   
SMALL HiBM       0.000623   3.200044  1.374067e-03   
ME2 BM1          0.000630  -2.737954  6.182275e-03   
ME2 BM2          0.000482   0.506436  6.125507e-01   
ME2 BM3          0.000533   1.357207  1.747153e-01   
ME2 BM4          0.000466   1.301431  1.931111e-01   
ME2 BM5          0.000510   0.528445  5.971902e-01   
ME3 BM1          0.000600  -1.556241  1.196507e-01   
ME3 BM2          0.000554   1.458669  1.446562e-01   
ME3 BM3          0.000576   0.066297  9.471410e-01   
ME3 BM4          0.000557   1.107741  2.679735e-01   
ME3 BM5          0.000712   0.598837  5.492817e-01   
ME4 BM1          0.000595   1.666856  9.554295e-02   
ME4 BM2          0.000642  -0.633446  5.264426e-01   
ME4 BM3          0.000661  -0.165968  8.681818e-01   
ME4 BM4          0.000688   0.637518  5.237874e-01   
ME4 BM5          0.000781  -0.641197  5.213949e-01   
BIG LoBM         0.000425   3.894440  9.842586e-05   
ME5 BM2          0.000550   0.113401  9.097125e-01   
ME5 BM3          0.000650   0.055580  9.556761e-01   
ME5 BM4          0.000668  -3.237419  1.206160e-03   
BIG HiBM         0.000986  -2.025729  4.279254e-02   

           Hypothesis at 5% FF3 P-val test Hypothesis at 5% FF3 T-test  
SMALL LoBM                     Reject Null                 Reject Null  
ME1 BM2                 Cannot Reject Null          Cannot Reject Null  
ME1 BM3                 Cannot Reject Null          Cannot Reject Null  
ME1 BM4                        Reject Null                 Reject Null  
SMALL HiBM                     Reject Null                 Reject Null  
ME2 BM1                        Reject Null                 Reject Null  
ME2 BM2                 Cannot Reject Null          Cannot Reject Null  
ME2 BM3                 Cannot Reject Null          Cannot Reject Null  
ME2 BM4                 Cannot Reject Null          Cannot Reject Null  
ME2 BM5                 Cannot Reject Null          Cannot Reject Null  
ME3 BM1                 Cannot Reject Null          Cannot Reject Null  
ME3 BM2                 Cannot Reject Null          Cannot Reject Null  
ME3 BM3                 Cannot Reject Null          Cannot Reject Null  
ME3 BM4                 Cannot Reject Null          Cannot Reject Null  
ME3 BM5                 Cannot Reject Null          Cannot Reject Null  
ME4 BM1                 Cannot Reject Null          Cannot Reject Null  
ME4 BM2                 Cannot Reject Null          Cannot Reject Null  
ME4 BM3                 Cannot Reject Null          Cannot Reject Null  
ME4 BM4                 Cannot Reject Null          Cannot Reject Null  
ME4 BM5                 Cannot Reject Null          Cannot Reject Null  
BIG LoBM                       Reject Null                 Reject Null  
ME5 BM2                 Cannot Reject Null          Cannot Reject Null  
ME5 BM3                 Cannot Reject Null          Cannot Reject Null  
ME5 BM4                        Reject Null                 Reject Null  
BIG HiBM                       Reject Null                 Reject Null

### Question 4)

The following is an output for the mean of the betas related to the FF3 model: As follows:
betas_FF3_mkt
betas_FF3_smb
betas_FF3_hml
We have different betas for each of the 25 portfolios, and so the average shows that difference.
The average beta for smb is at 0.54 and the average beta for hml is at 0.21 which shows a great gap on average.

In [ ]:
mean_of_betas(df_FF3)

1.0162309926811424
0.5421405314214436
0.2145237305349336


### Question 5

The CAPM suggest that the best performing portfolio is:
- SMALL HiBM with an alpha of :0.0053
- ME1 BM4 with an alpha of : 0.0041

CAPM worst:
- SMALL LoBM alpha : -0.0049

FF3 best:
- SMALL HiBM with an alpha of :0.002
- ME1 BM4 with an alpha of : 0.0016

FF3 worst:
- SMALL LoBM alpha : -0.0047

In both models, we can see that the better performing are small sized and medium book marke portfolios.



### Question 6)

If I were an asset manager, I would invest in SMALL HiBM and ME1 BM4, as both models suggest. And I would also short the worst performing portfolio which has a negative return, SMALL LoBM. By shorting this portfolio I would be able to invest more on the better performing portfolios.


# Part III: An Optimal Portfolio of Industries

## Modern Portfolio Theory

Modern Portfolio Theory (MPT) is an investment theory developed by Harry Markowitz and published under the title "Portfolio Selection" in the Journal of Finance in 1952.

There are a few underlying concepts that can help anyone to understand MPT. If you are familiar with finance, you might know what the acronym "TANSTAAFL" stands for. It is a famous acronym for "There Ain't No Such Thing As A Free Lunch". This concept is also closely related to 'risk-return trade-off'.

Higher risk is associated with greater probability of higher return and lower risk with a greater probability of smaller return. MPT assumes that investors are risk-averse, meaning that given two portfolios that offer the same expected return, investors will prefer the less risky one. Thus, an investor will take on increased risk only if compensated by higher expected returns.

Another factor comes in to play in MPT is "diversification". Modern portfolio theory says that it is not enough to look at the expected risk and return of one particular stock. By investing in more than one stock, an investor can reap the benefits of diversification – chief among them, a reduction in the riskiness of the portfolio.

What you need to understand is "risk of a portfolio is not equal to average/weighted-average of individual stocks in the portfolio". In terms of return, yes it is the average/weighted average of individual stock's returns, but that's not the case for risk. The risk is about how volatile the asset is, if you have more than one stock in your portfolio, then you have to take count of how these stocks movement correlates with each other. The beauty of diversification is that you can even get lower risk than a stock with the lowest risk in your portfolio, by optimising the allocation.

I will try to explain as I go along with the actual code. First, let's start by importing some libraries we need. "Quandl" is a financial platform which also offers Python library. If you haven't installed it before, of course, you first need to install the package in your command line "!pip -q install quandl", and before you can use it, you also need to get an API key on Quandl's website. Sign-up and getting an API key is free but has some limits. As a logged-in free user, you will be able to call 2,000 calls per 10 minutes maximum (speed limit), and 50,000 calls per day (volume limit). https://www.quandl.com/

### Construct excess returns on all industries
####Load Ken French's industry data
https://mba.tuck.dartmouth.edu/pages/faculty/ken.french/data_library.html

In [ ]:
import pandas as pd
import numpy as np

import pandas_datareader.data as web  # module for reading datasets directly from the web
p = web.DataReader('12_Industry_Portfolios', 'famafrench', start=1900)[0]/100



In [ ]:
p.head()

NoDur   Durbl   Manuf   Enrgy   Chems   BusEq   Telcm   Utils  \
Date                                                                      
1926-07  0.0145  0.1555  0.0367 -0.0118  0.0801  0.0316  0.0083  0.0704   
1926-08  0.0397  0.0368  0.0242  0.0347  0.0514  0.0197  0.0217 -0.0169   
1926-09  0.0114  0.0480 -0.0007 -0.0339  0.0530 -0.0034  0.0241  0.0204   
1926-10 -0.0124 -0.0823 -0.0316 -0.0078 -0.0455 -0.0538 -0.0011 -0.0263   
1926-11  0.0520 -0.0019  0.0382  0.0001  0.0511  0.0479  0.0163  0.0371   

          Shops   Hlth    Money   Other  
Date                                     
1926-07  0.0011  0.0177 -0.0002  0.0222  
1926-08 -0.0071  0.0425  0.0447  0.0434  
1926-09  0.0021  0.0069 -0.0161  0.0037  
1926-10 -0.0229 -0.0057 -0.0551 -0.0273  
1926-11  0.0643  0.0542  0.0234  0.0210

#### Load the risk-free rate from Ken French's Factors data

In [ ]:
p_factors = web.DataReader('F-F_Research_Data_Factors', 'famafrench', start=1900)[0]
rf = p_factors[['RF']]/100
rf.head()

RF
Date           
1926-07  0.0022
1926-08  0.0025
1926-09  0.0023
1926-10  0.0032
1926-11  0.0031

#### Compute excess returns

In [ ]:
rx = pd.concat([p, rf], axis=1)
for col in p.columns:
  rx[col] = rx[col] - rx.RF

rx = rx.drop('RF', axis=1)
rx.head()

NoDur   Durbl   Manuf   Enrgy   Chems   BusEq   Telcm   Utils  \
Date                                                                      
1926-07  0.0123  0.1533  0.0345 -0.0140  0.0779  0.0294  0.0061  0.0682   
1926-08  0.0372  0.0343  0.0217  0.0322  0.0489  0.0172  0.0192 -0.0194   
1926-09  0.0091  0.0457 -0.0030 -0.0362  0.0507 -0.0057  0.0218  0.0181   
1926-10 -0.0156 -0.0855 -0.0348 -0.0110 -0.0487 -0.0570 -0.0043 -0.0295   
1926-11  0.0489 -0.0050  0.0351 -0.0030  0.0480  0.0448  0.0132  0.0340   

          Shops   Hlth    Money   Other  
Date                                     
1926-07 -0.0011  0.0155 -0.0024  0.0200  
1926-08 -0.0096  0.0400  0.0422  0.0409  
1926-09 -0.0002  0.0046 -0.0184  0.0014  
1926-10 -0.0261 -0.0089 -0.0583 -0.0305  
1926-11  0.0612  0.0511  0.0203  0.0179

## Compute portfolio return statistics

Compute mean returns on all industries and their covariance matrix, as we did in class. This should require only calling two simple functions. Consult the *Efficient Frontier and Modern Portfolio Theory* notebook if unsure.

In [ ]:
mean_returns = rx.mean()
cov_matrix = rx.cov()

Print mean excess returns:

In [ ]:
mean_returns

NoDur    0.006942
Durbl    0.008983
Manuf    0.007642
Enrgy    0.007698
Chems    0.007516
BusEq    0.008734
Telcm    0.005664
Utils    0.006184
Shops    0.007550
Hlth     0.008079
Money    0.007423
Other    0.005740
dtype: float64

## Minimum variance portfolio
Compute the minimum variance portfolio using the formula in class. The easiest way to obtain this is to compute
$$w = V^{-1} \mathbf{1}_N,$$
where $\mathbf{1}_N$ is the column vector of ones and $V^{-1}$ is the inverse of the variance-covariance matrix of industry returns. Then, simply rescale all weights so that sum one to one (by dividing by their sum). This is equivalent to the formula I showed in class.


In [ ]:
y = np.linalg.inv(cov_matrix)@(np.ones((12,), dtype=int))
y = pd.Series(y / y.sum(), index=mean_returns.index)
print(y)

NoDur    0.659427
Durbl   -0.036664
Manuf   -0.216841
Enrgy    0.168347
Chems    0.100097
BusEq   -0.087873
Telcm    0.506926
Utils    0.169408
Shops    0.024945
Hlth     0.107469
Money   -0.411688
Other    0.016447
dtype: float64


## Maximum Sharpe Ratio portfolio
Compute the maximum Sharpe ratio portfolio using the formula in class. The easiest way to obtain this is to compute
$$w = V^{-1} \left( E[r] - r_f \right).$$
Then, simply rescale all weights so that sum one to one (by dividing by their sum). This is equivalent to the formula I showed in class.
*Hint:* I have shown you how to compute this in the *Efficient Frontier and Modern Portfolio Theory* notebook.

In [ ]:
w = np.linalg.inv(cov_matrix) @ (mean_returns)
w = pd.Series(w / w.sum(), index=mean_returns.index)
print(w)

NoDur    0.750613
Durbl    0.118130
Manuf   -0.057490
Enrgy    0.288203
Chems    0.024734
BusEq    0.130730
Telcm    0.250072
Utils    0.055885
Shops    0.044834
Hlth     0.313418
Money   -0.278927
Other   -0.640201
dtype: float64


# Answers Part III: An Optimal Portfolio of Industries (35 points)

### Question 1

From the following output we can undertand that most industries have performed well. The one that have stood out are the following:

-Durbl    0.008983

-BusEq    0.008734

-Hlth     0.008079

In [ ]:
mean_returns

NoDur    0.006942
Durbl    0.008983
Manuf    0.007642
Enrgy    0.007698
Chems    0.007516
BusEq    0.008734
Telcm    0.005664
Utils    0.006184
Shops    0.007550
Hlth     0.008079
Money    0.007423
Other    0.005740
dtype: float64

### Question 2


In the following out we have the investment weights into the different sectors.
The highest weight is given to:
- NoDur    0.659427
And lowest weight is given to:
- Money   -0.411688

Our minimum variance portfolio suggest that we short BusEq at a weight of
-0.087873 yet it was one of the industries with the highest average returns with 0.008734. This is because the minimum variance portfolio takes into account the risk related to one portfolio and so with the amount of risk in BusEq our formula found that in order to have a minimum variance (risk) portfolio we would need to short this industry.

In [ ]:
y = np.linalg.inv(cov_matrix)@(np.ones((12,), dtype=int))
y = pd.Series(y / y.sum(), index=mean_returns.index)
print(y)

NoDur    0.659427
Durbl   -0.036664
Manuf   -0.216841
Enrgy    0.168347
Chems    0.100097
BusEq   -0.087873
Telcm    0.506926
Utils    0.169408
Shops    0.024945
Hlth     0.107469
Money   -0.411688
Other    0.016447
dtype: float64


### Question 3

In the following output we can see that:
Highest weight:
- NoDur    0.750613
Lowest weight:
- Money   -0.278927

The weights given by the minimum variance portfolio and the highest sharpe ratio portfolio differ for some industries.

The maximum sharpe ratio portfolio takes into account the highest possible expected excess return per unit of risk. And so, it would possibly short some industries so taht it may increase the investement in other well defined industries for this case.

We can see that Other and Money have been highly shorted so that Hlth and NoDur can increase in weight of investement.



In [ ]:
w = np.linalg.inv(cov_matrix) @ (mean_returns)
w = pd.Series(w / w.sum(), index=mean_returns.index)
print(w)

NoDur    0.750613
Durbl    0.118130
Manuf   -0.057490
Enrgy    0.288203
Chems    0.024734
BusEq    0.130730
Telcm    0.250072
Utils    0.055885
Shops    0.044834
Hlth     0.313418
Money   -0.278927
Other   -0.640201
dtype: float64
